Updates to quiz/lesson resource selection metadata display

rtibbles opened the pull request:
## Summary
* Updates the channel API to return `included_categories` and `included_grade_levels` in array form, parallel to how similar fields are returned from contentnode endpoints
* Adds handling for a previously uncaught bad string mapping for "Professional" in the context of the grade_levels metadata
* Reverts displayed counts in quiz resource selection for channels and folders to be for unused questions, consistent with exercises
* Makes positioning of 'count type' metadata consistent across card types (always displayed above the description if it exists)
* Enhances the content node API endpoint to return the already annotated `on_device_resources` count
* Generalizes our handling of fields that get added to the public contentnode API endpoint to systematically handle missing keys from older versions of the API (without having to fully version the API)
* Labels folders with resource counts in lesson resource selection
* Updates annotation of the Bookmarks top level card and individual bookmark cards to show unused questions when in quiz selection mode
* Updates the SelectFromQuizSearchResults component to reuse the QuizSelectionHeader when not selecting a practice quiz, parallel to the logic in SelectFromTopics.

## References
Fixes #13240 
Fixes #13255

## Reviewer guidance
Note that I went a little bit above and beyond by adding the folder counts here on lessons - but it also means we if we decide to standardize quizzes on this too, we're ready to go.

| Description | Before | After |
|---------- |----------|--------|
| Bookmarks and Channel counts for quizzes| ![Screenshot from 2025-03-26 13-06-01](https://github.com/user-attachments/assets/1e4a5c39-cbd0-4ad0-97c8-ab8704a53189) | ![Screenshot from 2025-03-26 12-33-04](https://github.com/user-attachments/assets/be7a0d1e-0260-478d-9195-95bba9a910b7) |
| Bookmark card annotation for quizzes | ![Screenshot from 2025-03-26 13-06-05](https://github.com/user-attachments/assets/cc29b747-9360-48fd-8ff3-033c637ab2e1) | ![Screenshot from 2025-03-26 12-33-08](https://github.com/user-attachments/assets/d2e274e7-ab65-4dad-8a53-03e851762a07) |
| Folder card annotation for quizzes | ![Screenshot from 2025-03-26 13-06-28](https://github.com/user-attachments/assets/f6fb6692-6082-4dd9-a2f0-8cbec954fc1c) | ![Screenshot from 2025-03-26 12-33-14](https://github.com/user-attachments/assets/bc24c38b-a381-4251-99b2-a9edf0b0326f) |
| Header for quiz search | ![Screenshot from 2025-03-26 13-06-50](https://github.com/user-attachments/assets/34d96cbc-e1f5-457f-b34d-c74e213cb2c2) | ![Screenshot from 2025-03-26 12-33-31](https://github.com/user-attachments/assets/bb8c3d6f-1d26-425d-9e39-7607dd405162) |
| Folder card annotation for lessons | ![Screenshot from 2025-03-26 13-07-13](https://github.com/user-attachments/assets/8640c02e-fc60-4595-a5f9-c1ebe19f7f73) | ![Screenshot from 2025-03-26 12-33-53](https://github.com/user-attachments/assets/9a3846dc-f051-4ed2-9b19-20980496099b)
 |

rtibbles committed:
Update channel API to return fields consistent with the content node API.

rtibbles committed:
Properly handle string translation for grade_level "Professional".

rtibbles committed:
Make card resource/question count labelling consistent across card types.
Reinstate folder/channel unused counts.

rtibbles committed:
Add descendant resource counts to folders in lesson preview.
Clean up unused descendants endpoint.

rtibbles committed:
Update message display on bookmarks to show unused questions in quiz selection.

rtibbles committed:
Use QuizSelectionHeader when relevant.

marcellamaki commented on kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/sidePanels/QuizResourceSelection/index.vue:
```diff
@@ -364,15 +365,31 @@
-          .filter(questionItem => !workingQuestions.value.some(q => q.item === questionItem));
-        return questionsItemsUnused.length;
-      };
+      const unusedQuestionsCount = useMemoize(content => {
```
using useMemoize here is pretty smart

rtibbles replied:
I stole it from our 0.17 implementation: https://github.com/learningequality/kolibri/blame/release-v0.17.x/kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue#L492

marcellamaki commented on kolibri/core/content/api.py:
```diff
@@ -621,6 +621,15 @@ def map_file(file):
+# A map of fields that did not used to be in the BaseContentNodeMixin
+# but now are - the map gives a default value for these to be filled in with
+# if not present on an API request from a contentnode public API endpoint
+contentnode_previously_omitted_fields = {
```
So, we can continue to update this map, as long as we are also updating the fields below? And we combined the previously omitted field learner needs here, and then adding the on_device_resources_field on line 657?

rtibbles replied:
Yes - this will allow us to continue adding to the contentnode public API, but make sure that any remote requests to it made using the `baseurl` query parameter will return the correct shape response even if it's an older version.

marcellamaki commented on kolibri/plugins/coach/assets/src/views/common/resourceSelection/subPages/SelectFromBookmarks.vue:
```diff
@@ -79,7 +79,11 @@
       const { data, hasMore, fetchMore, loadingMore } = props.bookmarksFetch;
 
-      const contentCardMessage = content => {
+      const wrappedContentCardMessage = content => {
```
This is maybe a nitpick, but I really don't understand what the `wrapped` is conveying here.

rtibbles replied:
It's because `contentCardMessage` is already the name of a prop, so it gives a name collision if I exported it with the same name.

As it is wrapping the `contentCardMessage` prop (it invokes it first), I thought `wrappedContentCardMessage` was as good a name as any.

marcellamaki replied:
fair enough!

marcellamaki reviewed (commented):
Hi @rtibbles - I've just done an intial code read through and added a few questions for clarification. I haven't checked it out yet and done any manual QA. Maybe @pcenov can take a look tomorrow (Thursday), and I'll re-review morning my time as well.

pcenov commented:
Hi @rtibbles,

1. In `Lessons > Manage resources > Search > Search results` we are not showing the number of resources for the first topic of the topic tree - should we be showing those in the search results in the first place?

https://github.com/user-attachments/assets/8ceb1aaf-861d-4b57-bfed-3561341bbc8e

2. In `Quizzes > Select quiz` what should be the behavior when selecting a practice quiz? Currently I am seeing the Bookmarks section (in which I don't have any practice quizzes) and I see '0 questions unused' plus there's no question count on the other folders:

https://github.com/user-attachments/assets/48b8b405-f651-478d-a68b-63643e1a5c7c

rtibbles committed:
Ensure we properly label lesson resource selection search results.
Pass through disabled to ensure we properly disable items due to loading states.

rtibbles committed:
When reannotating bookmarks make sure to update count.
Be consistent and don't show unused question counts on bookmarks when selecting practice quizzes.

rtibbles commented:
Hi @pcenov

1. Is completely addressed.
![Screenshot from 2025-03-27 08-09-31](https://github.com/user-attachments/assets/6757728f-36ae-497d-985d-c36b7bf826c8)

2. In consultation with @marcellamaki I have addressed in this way:
* ensure that we do not show the Bookmarks at all when there are no practice quizzes in bookmarks
* continue to not show either unused question or resource counts when selecting practice quizzes (the former because it is misleading for the 'select practice quiz flow', the latter because it is technically difficult to get these numbers at present
![Screenshot from 2025-03-27 08-35-44](https://github.com/user-attachments/assets/e10e9e54-41c1-46d1-9e6c-f0e4bf70256c)

nucleogenesis commented on kolibri/plugins/coach/assets/src/views/common/resourceSelection/subPages/SelectionIndex.vue:
```diff
@@ -98,7 +98,16 @@
           :headingLevel="3"
-        />
+        >
+          <template #belowTitle>
+            <div
+              v-if="contentCardMessage(channel)"
+              style="margin: 0 0 0.5rem"
+            >
+              {{ contentCardMessage(channel) }}
+            </div>
+          </template>
```
This text was previously displayed through a KTextTruncator with maxlines=1 - was it intentional not to carry that over here?

rtibbles replied:
It was deliberate inasmuch as it was intended to be consistent with how this text was displayed on folder and resource cards - but perhaps they should all be following the truncator pattern instead?

marcellamaki replied:
I do think being consistent across the cards would be useful. I don't know if I have a strong thought about using the truncator or not. I guess one of the benefits would be consistent spacing, and we probably wouldn't be losing that much information

marcellamaki replied:
Honestly I don't have too strong of an opinion either pro or against truncation, but I would be in support of consistency across the cards

rtibbles replied:
I think simply because we are keenly aware that some translated strings do end up being longer than others, it is probably worth being proactive with the truncator, and we can apply that everywhere we are inserting card messages.

rtibbles replied:
I have updated every instance of text injected in the card to be wrapped in KTextTruncator.

I also made the inline styling of all the descriptions and KTextTruncators consistent, and ensured we were using KDS mandated spacing (8px) instead of rem/em units: https://design-system.learningequality.org/layout#spacing

I tested on a smaller screen size and I didn't observe any differences when I tried with 4px for the smaller screen sizes, so I left it as is (possibly due to the min-height).

rtibbles committed:
Use KTextTruncator for all inserted card messages.
Use consistent styling for messages and descriptions across card types.
Use KDS 8px spacing rather than relative units.
Add br to ensure new line before description.

rtibbles commented:
No idea why, but readthedocs seems to be failing to checkout the revision here - I don't think this is a genuine issue with this PR though.

pcenov commented:
Hi @rtibbles - I confirm that the issues reported by me are resolved now.

marcellamaki reviewed (approved):
Thanks @rtibbles, @pcenov , and @nucleogenesis !
